### ISSUE TITLE
Conditional MaskedAutoregressive Flow outputs NAN

### ISSUE THREAD

**Issue by JeffTheHacker** [tensorflow/probability#1203]
Hi, I am experimenting with the bijector (MaskedAutoregressive Flow). Essentially I want it to be a mapping from a length 20 vector drawn from some base distribution to some other distribution, conditioned on some input value (which is a length 10 vector). Below is my code.

`
import tensorflow_probability as tfp
import tensorflow as tf
import numpy as np
tfd = tfp.distributions
tfb = tfp.bijectors

input = np.ones(shape=(20)).astype(np.float32)
condition = np.random.normal(size=(10)).astype(np.float32) + 10

fn = tfb.AutoregressiveNetwork(params=2, event_shape=20, conditional=True, conditional_event_shape=10, hidden_units=[10, 10])
bijector = tfb.MaskedAutoregressiveFlow(fn)

print(bijector.forward(input, conditional_input=condition))`

Result of running the code gives me:

`tf.Tensor(
[nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan], shape=(20,), dtype=float32)`

I don't think my input or condition vectors are big enough to cause exploding values. Any ideas on what I might be doing wrong? Thanks!

For reference, I am using tensorflow-probability = 0.11.0

**Comment by JeffTheHacker** [comment]
I looked into the code segment, and apparently the issue is caused by the underlying autoregressive network: it doesn't allow setting activation functions for conditional inputs. In particular I'm talking about the "conditional_output" below. 
```

def build(self, input_shape):
    """See tfkl.Layer.build."""
    if self._event_shape is None:
      # `event_shape` wasn't specied at __init__, so infer from `input_shape`.
      self._event_shape = [tf.compat.dimension_value(input_shape[-1])]
      self._event_size = self._event_shape[-1]
      self._event_ndims = len(self._event_shape)
      # Should we throw if input_shape has rank > 2?

    if input_shape[-1] != self._event_shape[-1]:
      raise ValueError('Invalid final dimension of `input_shape`. '
                       'Expected `{!r}`, but got `{!r}`'.format(
                           self._event_shape[-1], input_shape[-1]))

    # Construct the masks.
    self._input_order = _create_input_order(
        self._event_size,
        self._input_order_param,
    )
    self._masks = _make_dense_autoregressive_masks(
        params=self._params,
        event_size=self._event_size,
        hidden_units=self._hidden_units,
        input_order=self._input_order,
        hidden_degrees=self._hidden_degrees,
    )

    outputs = [tf.keras.Input((self._event_size,), dtype=self.dtype)]
    inputs = outputs[0]
    if self._conditional:
      conditional_input = tf.keras.Input((self._conditional_size,),
                                         dtype=self.dtype)
      inputs = [inputs, conditional_input]

    # Input-to-hidden, hidden-to-hidden, and hidden-to-output layers:
    #  [..., self._event_size] -> [..., self._hidden_units[0]].
    #  [..., self._hidden_units[k-1]] -> [..., self._hidden_units[k]].
    #  [..., self._hidden_units[-1]] -> [..., event_size * self._params].
    layer_output_sizes = self._hidden_units + [self._event_size * self._params]
    for k in range(len(self._masks)):
      autoregressive_output = tf.keras.layers.Dense(
          layer_output_sizes[k],
          activation=None,
          use_bias=self._use_bias,
          kernel_initializer=_make_masked_initializer(
              self._masks[k], self._kernel_initializer),
          bias_initializer=self._bias_initializer,
          kernel_regularizer=self._kernel_regularizer,
          bias_regularizer=self._bias_regularizer,
          kernel_constraint=_make_masked_constraint(
              self._masks[k], self._kernel_constraint),
          bias_constraint=self._bias_constraint,
          dtype=self.dtype)(outputs[-1])
      if (self._conditional and
          ((self._conditional_layers == 'all_layers') or
           ((self._conditional_layers == 'first_layer') and (k == 0)))):
        conditional_output = tf.keras.layers.Dense(
            layer_output_sizes[k],
            activation=None,
            use_bias=False,
            kernel_initializer=self._kernel_initializer,
            bias_initializer=None,
            kernel_regularizer=self._kernel_regularizer,
            bias_regularizer=None,
            kernel_constraint=self._kernel_constraint,
            bias_constraint=None,
            dtype=self.dtype)(conditional_input)
        outputs.append(tf.keras.layers.Add()([
            autoregressive_output,
            conditional_output]))
      else:
        outputs.append(autoregressive_output)
      if k + 1 < len(self._masks):
        outputs.append(
            tf.keras.layers.Activation(self._activation)
            (outputs[-1]))
    self._network = tf.keras.models.Model(
        inputs=inputs,
        outputs=outputs[-1])
    # Record that the layer has been built.
    super(AutoregressiveNetwork, self).build(input_shape)
```

**Comment by SiegeLordEx** [comment]
Neural net initializers tend to assume that the network inputs have elements with magnitude <= 1 and a mean of 0 (across the set of possible inputs). The elements of your `condition` variable have magnitude of `10`, however, so one could expect something to break, especially with the default affine pointwise transformation `tfb.MaskedAutoregressiveFlow` uses. I'd consider the following options:

1. Normalize your condition values to be smaller, both in mean and magnitude.
2. Adjust the `kernel_initializer` of the entire AR network to be smaller (good idea anyway, these things train best when they're initialized to something close to an identity transformation).
3. We (TFP devs) probably could provide a separate kernel initializer just for the linear condiitonal transformation, so you could control the scaling separately.

As for the lack of configurability of the conditional tranformation activation function, that's pretty typical as far as conditioning in neural nets goes: typically the conditional inputs are just added onto the pre-activation outputs of layers.

**Comment by chayanchatterjee** [comment]
I have a related question. I am using a conditional MAF with 3 MADE blocks to approximate a target distribution from a multivariate Gaussian. My conditional values are within -3 and 1 and has mean 0 and std 1. I notice that when I sample from the trained model and plot the samples, the points have the desired shape of the target distribution, but when I evaluate the probability density over a meshgrid, the high density regions lie away from the cluster of points drawn from the model, and also the overall region of high density is spread out over a much larger area of the meshgrid than the region the cluster of points lie in. I have added some relevant parts of my code below. Any help would be appreciated.

      
            x_ = tf.keras.layers.Input(shape=self.y_train.shape[-1], dtype=tf.float32)
            self.encoded_features = ResNet34(input1, input2, self.filters_real, self.kernel_size, self.strides, self.pool_size, 
    self.prev_filters_real, input_shapes1=[2, self.n_samples, self.n_det], input_shapes2 = 
    self.intrinsic_train.shape[-1]).construct_model()
            bijectors = []
            for i in range(self.num_bijectors):
                masked_auto_i = self.make_masked_autoregressive_flow(i, hidden_units = self.MAF_hidden_units, activation = 'relu', conditional_event_shape=self.encoded_features.shape[-1])
                
                bijectors.append(masked_auto_i)
    
                USE_BATCHNORM = True
    
                if USE_BATCHNORM:
                # BatchNorm helps to stabilize deep normalizing flows, esp. Real-NVP
                    bijectors.append(tfb.BatchNormalization(name='batch_normalization'+str(i)))

                bijectors.append(tfb.Permute(permutation = [2, 1, 0]))
                
                flow_bijector = tfb.Chain(list(reversed(bijectors[:-1])))
            
            # Define the trainable distribution
            self.trainable_distribution = tfd.TransformedDistribution(distribution=tfd.Sample(tfd.Normal(loc=0, scale=1),sample_shape=[3]), bijector = flow_bijector)
            
            
            log_prob_ = self.trainable_distribution.log_prob(x_, bijector_kwargs=
                        self.make_bijector_kwargs(self.trainable_distribution.bijector, 
                                             {'maf.': {'conditional_input':self.encoded_features}}))

            self.model = tf.keras.Model([input1, input2, x_], log_prob_)
            self.encoder = tf.keras.Model([input1, input2], self.encoded_features) 

            opt = tf.keras.optimizers.Adam(learning_rate=self.lr)  # optimizer
            checkpoint = tf.train.Checkpoint(optimizer=opt, model=self.model)
            


        self.train(log_prob_, checkpoint)
    
    

    # Define the trainable distribution
    def make_masked_autoregressive_flow(self, index, hidden_units, activation, conditional_event_shape):
        
        made = tfp.bijectors.AutoregressiveNetwork(params=2,
              hidden_units=hidden_units,
              event_shape=(3,),
              activation=activation,
              conditional=True,
              kernel_initializer = tf.keras.initializers.RandomNormal(),
              conditional_event_shape=conditional_event_shape)
    
        return tfp.bijectors.MaskedAutoregressiveFlow(shift_and_log_scale_fn = made, name='maf'+str(index))

    def make_bijector_kwargs(self, bijector, name_to_kwargs):
        
        if hasattr(bijector, 'bijectors'):
            
            return {b.name: self.make_bijector_kwargs(b, name_to_kwargs) for b in bijector.bijectors}
    
        else:
            for name_regex, kwargs in name_to_kwargs.items():
                if re.match(name_regex, bijector.name):
                    return kwargs
        return {}
           
   
    def train(self, log_prob, checkpoint):
        """Compiles and trains the model"""
        
        self.model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=self.lr), loss=lambda _, log_prob: -tf.reduce_mean(log_prob))
        model_history = self.model.fit([self.X_train, self.intrinsic_train, self.y_train], np.zeros((len(self.X_train_real)), dtype=np.float32),
              batch_size=self.batch_size,
              epochs=self.epochs,
              validation_split=self.val_split,
              callbacks=callbacks_list,
              shuffle=True,
              verbose=True)

     x_test = np.expand_dims(self.X_test[i], axis=0)
     intrinsic_test = np.expand_dims(self.intrinsic_test[i], axis=0)
    
     preds = self.encoder([x_test, intrinsic_test])
     prob_density = tf.exp(self.trainable_distribution.log_prob(pixels, bijector_kwargs=self.make_bijector_kwargs(self.trainable_distribution.bijector, {'maf.': {'conditional_input':preds}})))
            
     samples = self.trainable_distribution.sample((2000,), bijector_kwargs=self.make_bijector_kwargs(self.trainable_distribution.bijector, {'maf.': {'conditional_input':preds}}))
